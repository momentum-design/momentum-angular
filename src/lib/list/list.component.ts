/** @component list */

import {
  Component,
  OnInit,
  Input,
  HostBinding,
  QueryList,
  ContentChildren,
  ChangeDetectorRef,
  AfterContentInit,
  Output,
  EventEmitter,
  NgZone,
  ElementRef
} from '@angular/core';
import { uniqueId } from 'lodash';
import { ListItemComponent, OptionSelectionChange } from '../list-item';
import { Observable, defer, merge } from 'rxjs';
import { take, switchMap } from 'rxjs/operators';

@Component({
  selector: 'md-list',
  template: `
    <ng-content></ng-content>
  `,
  styles: [],
  host: {
    'class': 'md-list',
    '[class.md-list--wrap]': 'wrap'
  }
})
export class ListComponent implements OnInit, AfterContentInit {

  private Subscription;

  constructor(
    private cd: ChangeDetectorRef,
    private _ngZone: NgZone,
    private el: ElementRef) {

    this.optionSelectionChanges.subscribe(event => {
      this._onSelect(event.source);
    });
  }

  /** @option Optional ID value of List | null */
  @HostBinding('id') @Input() id: string = uniqueId('md-list-');
  /** @option itemRole prop to set child roles | 'listItem' */
  @Input() itemRole = 'listItem';
  /** @option Callback function invoked by user selecting an interactive item within list | null */
  @Output() select = new EventEmitter();
  /** @option Sets the ARIA role for the Nav, in the context of a TabContainer | 'list' */
  @HostBinding('attr.role') @Input() role: string = 'list';
  /** @option Sets the orientation of the List | 'vertical' */
  private _tabType: string = 'vertical';
  @Input()
  set tabType(tabType: string) {
    if (this._tabType) {
      this.el.nativeElement.classList.remove(`md-list--${this._tabType}`);
    }
    this.el.nativeElement.classList.add(`md-list--${tabType}`);
    this._tabType = tabType;
  }
  /** @option Type sets List size | null */
  @Input() type = null;
  /** @option Wrap prop type to wrap items to next row */
  @Input() wrap = false;

  @ContentChildren(ListItemComponent) listItems: QueryList<ListItemComponent>;

  /** Combined stream of all of the child options' change events. */
  readonly optionSelectionChanges: Observable<OptionSelectionChange> = defer(() => {
    if (this.listItems) {
      return merge(...this.listItems.map(option => option.selectionChange));
    }

    return this._ngZone.onStable.asObservable()
      .pipe(take(1), switchMap(() => this.optionSelectionChanges));
  }) as Observable<OptionSelectionChange>;

  ngOnInit() {
    if (this._tabType && !this._isTabTypeOptionValid()) {
      throw new Error(`md-list: List tabType option must be one of the following:
        vertical, horizontal`);
    }
    if (this.type && !this._isTypeOptionValid()) {
      throw new Error(`md-list: List type option must be one of the following:
        small, large, space, xlarge`);
    }
  }

  ngAfterContentInit() {
    // set up listItems children with props
    this.listItems.forEach(listItem => {
      if (this.type) {
        listItem.type = this.type;
      }
      listItem.role = this.itemRole;
    });

    this.cd.detectChanges();
  }

  /** Invoked when an option is clicked. */
  private _onSelect(option: ListItemComponent): void {
    if (this.select) {
      this.listItems.forEach(listItem => {
        if (listItem !== option) {
          listItem.active = false;
        }
      });
      option.active = true;
      return this.select.emit(option);
    }
  }

  private _isTabTypeOptionValid = () => (
    ['vertical', 'horizontal'].includes(this._tabType)
  )

  private _isTypeOptionValid = () => (
    ['small', 'large', 'space', 'xlarge'].includes(this.type)
  )
}
