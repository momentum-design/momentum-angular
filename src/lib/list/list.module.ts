import { NgModule, CUSTOM_ELEMENTS_SCHEMA, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ListComponent } from './list.component';

@NgModule({
  imports: [CommonModule],
  declarations: [ListComponent],
  exports: [ListComponent],
  schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class ListModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: ListModule,
      providers: [],
    };
  }
}
