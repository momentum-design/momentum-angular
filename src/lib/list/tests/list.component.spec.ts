import { Component } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { ListComponent } from '../list.component';
import {
  IconModule,
  ListItemModule,
  ListItemSectionModule,
} from 'src/lib/public_api';

@Component({
  template: `
    <md-list
      [ngClass]="'custom-ng-class'"
      class="custom-class"
      id="1234"
      tabType="horizontal"
      type="small"
      itemRole="my-list-item"
      role="list-box"
      [wrap]="true"
      (select)="onSelect($event)"
      >
      <div md-list-item label='List Item A'></div>
      <div md-list-item label='List Item B'></div>
    </md-list>
  `
})
class TestHostComponent {}

describe('ListComponent', () => {
  let fixture: ComponentFixture<TestHostComponent>;
  let listComponent: ListComponent;
  let listNativeElement: HTMLElement;
  let testHost: TestHostComponent;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [ ListComponent, TestHostComponent ],
      imports: [
        IconModule,
        ListItemModule,
        ListItemSectionModule
      ]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestHostComponent);
    testHost = fixture.componentInstance;
    listComponent = fixture.debugElement.children[0].componentInstance;
    fixture.detectChanges();
  });

  it('should create and match snapshot', () => {
    fixture.detectChanges();
    expect(testHost).toBeTruthy();
    expect(listComponent).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should render with default class names in addition to generated class names from props', () => {
    fixture.detectChanges();

    listNativeElement = fixture.nativeElement;
    const list = listNativeElement.querySelector('md-list');
    expect(list).not.toBeNull();
    expect(list.className).toContain('custom-ng-class');
    expect(list.className).toContain('custom-ng-class');
    expect(list.className).toContain('custom-class');
    expect(list.className).toContain('md-list--wrap');
    expect(list.className).toContain('md-list--horizontal');
  });
});
