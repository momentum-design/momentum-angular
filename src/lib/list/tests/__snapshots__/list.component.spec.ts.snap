// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListComponent should create and match snapshot 1`] = `
<cui-list
  active="null"
  cd={[Function ViewRef_]}
  class=""
  focusFirst={[Function Boolean]}
  handleClick={[Function Function]}
  id={[Function String]}
  isActive={[Function Boolean]}
  isMulti="false"
  isTabTypeOptionValid={[Function Function]}
  isTypeOptionValid={[Function Function]}
  itemRole={[Function String]}
  listItems={[Function QueryList]}
  onSelect=""
  role={[Function String]}
  tabType={[Function String]}
  type="null"
  wrap="false"
>
  
</cui-list>
`;
